### PR TITLE
visual ameliorations

### PR DESCRIPTION
**Pull request by JaneBellaiche**
[ch-63596]

### CODE REVIEW

**Review by alexcombessie — CHANGES_REQUESTED** [review]
Not sure why but it fails on my local test project (DSS 9.0.1, python 3.7)
[dss-job-diag-DEVONTOLOGYTAGGING-Build_tagged__NP__2021-04-28T17-24-45.826_.zip](https://github.com/dataiku/dss-plugin-nlp-analysis/files/6393679/dss-job-diag-DEVONTOLOGYTAGGING-Build_tagged__NP__2021-04-28T17-24-45.826_.zip)

**Comment by alexcombessie** [comment]
Also, there is a conflict with the main branch. Can you merge and resolve it?

**Comment by JaneBellaiche** [comment]
Everything should work fine now

**Review by alexcombessie — CHANGES_REQUESTED** [review]
Good job! I have verified all items in the CH card.

I found a few UX nitpicks which were not in the card. Some of them are code suggestions inline.

Other suggestions:
 - rename "1 row / document (array column)" to "1 row / doc (array columns)" - and "1 row / document (nested JSON)" to "1 row / doc (nested JSON)
 - In array format with categories, the tag_list_{category} columns are missing descriptions. I would suggest "List of tags of category '{...}'"

File: python-lib/formatter_instanciator.py [hunk 1]
```diff
@@ -9,13 +9,13 @@
 # names of all additional columns depending on the output_format
 COLUMNS_DESCRIPTION = {
     "tag_keywords": "Matched keywords",
-    "tag_sentences": "Sentences with keywords",
-    "tag_json_full": "Detailed tag column",
-    "tag_json_categories": "category → tags",
-    "tag_list": "List of all found tags",
-    "tag": "Found tag",
+    "tag_sentences": "Sentences",
+    "tag_json_full": "Detailed tag column: count of occurrences, matched sentences and list of keywords per tag and category",
```
Comment (alexcombessie) [inline python-lib/formatter_instanciator.py:13]:
```suggestion
    "tag_json_full": "Detailed tag column: list of matched keywords per tag and category, count of occurrences, sentences containing matched keywords",
```

**Comment by JaneBellaiche** [comment]
I've commit the last changes you asked for :) 
Thank you for your feedback

File: python-lib/ontology_tagging_formatting.py [hunk 2]
```diff
@@ -1,7 +1,3 @@
-"""
-Module to write the output dataframe depending on the output format
```
Comment (alexcombessie) [inline python-lib/ontology_tagging_formatting.py:2]: Why remove that?

File: python-lib/ontology_tagging_formatting.py [hunk 3]
```diff
@@ -392,6 +406,10 @@ def _merge_df_columns_category(
         """
         output_df_copy = self.output_df.copy().add_prefix("tag_list_")
         tag_list_columns = output_df_copy.columns.tolist()
+        for column in tag_list_columns:
+            self.column_descriptions[column] = (
+                "List of tags for category " + column.split("_")[-1]
```
Comment (alexcombessie) [inline python-lib/ontology_tagging_formatting.py:411]: Use f"string formatting"

File: python-lib/ontology_tagger.py [hunk 4]
```diff
@@ -168,8 +170,13 @@ def _format_with_category(
...
         )
-        return formatter.write_df_category(
-            input_df=text_df, text_column=text_column, language_column=language_column
+        return (
+            formatter.write_df_category(
+                input_df=text_df,
+                text_column=text_column,
+                language_column=language_column,
+            ),
+            formatter.column_descriptions,
```
Comment (alexcombessie) [inline python-lib/ontology_tagger.py:179]: You need to update the type hints as the function signature changed

File: python-lib/ontology_tagger.py [hunk 5]
```diff
@@ -57,7 +57,9 @@ def __init__(
             enable_pipe_components="sentencizer",
         )
         self._matcher_dict = {}  # this will be filled by the _match_no_category method
-        self._keyword_to_tag = {}  # this will be filled by the _tokenize_keywords method
+        self._keyword_to_tag = (
```
Comment (alexcombessie) [inline python-lib/ontology_tagger.py:60]: Not sure why this was changed. Can you check your VScode formatter settings?

File: tests/python/unit/test_ontology_tagger.py [hunk 6]
```diff
@@ -87,7 +87,7 @@ def test_matching_in_lowercase():
         language="en",
         normalize_case=True,
     )
-    df = tagger.tag_and_format(
+    df, _ = tagger.tag_and_format(
```
Comment (alexcombessie) [inline tests/python/unit/test_ontology_tagger.py:90]: I am not a big fan of changing signatures of methods like that. I would prefer to keep the signature of tag_and_format, and introduce a new method to "compute_column_descriptions".

**Review by alexcombessie — CHANGES_REQUESTED** [review]
Need to make a dedicated method to compute column descriptions dynamically (and make sure column names are unique) cf https://github.com/dataiku/dss-plugin-nlp-preparation/blob/ccb40c9afa54b79269ffde219f18e5d6db1bfcfc/python-lib/text_cleaner.py#L86

File: python-lib/ontology_tagging_formatting.py (hunk 2 above)
Comment (JaneBellaiche) [reply python-lib/ontology_tagging_formatting.py:2]: I didn't, I moved it under the class declaration

File: python-lib/ontology_tagging_formatting.py (hunk 3 above)
Comment (JaneBellaiche) [reply python-lib/ontology_tagging_formatting.py]: 8565dce

File: python-lib/ontology_tagging_formatting.py [hunk 7]
```diff
@@ -376,16 +415,30 @@ def _merge_df_columns_category(
...
+            df=self.output_df,
+            columns=normalize_text(tag_list_columns),
+            prefix="tag_list",
         )
-        output_df_copy.insert(
-            len(self.output_df.columns), self.tag_columns[0], self.tag_sentences, True
+        self.output_df.columns = tag_list_columns_unique
+        for column, column_unique in zip(tag_list_columns, tag_list_columns_unique):
+            category = column.split("_")[-1]
+            self._column_descriptions[column_unique] = f"List of tags for category {category}"
```
Comment (alexcombessie) [inline python-lib/ontology_tagging_formatting.py:427]:
```suggestion
            self._column_descriptions[column_unique] = f"List of tags for category '{category}'"
```

File: python-lib/ontology_tagging_formatting.py [hunk 8]
```diff
@@ -376,16 +415,30 @@ def _merge_df_columns_category(
...
+            df=self.output_df,
+            columns=normalize_text(tag_list_columns),
+            prefix="tag_list",
         )
-        output_df_copy.insert(
-            len(self.output_df.columns), self.tag_columns[0], self.tag_sentences, True
+        self.output_df.columns = tag_list_columns_unique
+        for column, column_unique in zip(tag_list_columns, tag_list_columns_unique):
+            category = column.split("_")[-1]
+            self._column_descriptions[column_unique] = f"List of tags for category '{category}'"
```
Comment (alexcombessie) [inline python-lib/ontology_tagging_formatting.py:427]:
```suggestion
            self._column_descriptions[column_unique] = f"List of '{category}' tags"
```

**Review by alexcombessie — APPROVED** [review]

**Pull request merged**